lib/uplink/access.go

```diff
@@ -0,0 +1,112 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package uplink
+
+import (
+	"context"
+	"errors"
+
+	"storj.io/storj/pkg/storj"
+)
+
+// Access is all of the access information an application needs to store and
+// retrieve data. Someone with a share may have no restrictions within a project
+// (can create buckets, list buckets, list files, upload files, delete files,
+// etc), may be restricted to a single bucket, may be restricted to a prefix
+// within a bucket, or may even be restricted to a single file within a bucket.
+type Access struct {
+	Permissions Permissions
+	Uplink      *Uplink
+}
+
+// A Macaroon represents an access credential to certain resources
+type Macaroon interface {
```

Comment:
to be honest, let's rename this interface to be APIKey for now and get rid of the Restrict method and the Caveat type.